app/models/permissions/scientist_permission.rb

```diff
@@ -8,11 +8,40 @@ class ScientistPermission < BasePermission
     def initialize(user)
       super
       allow :scans, [:create]
-      # Check if Scientists should be able to upload labware
       allow :upload_labware, [:create]
+      allow :move_locations, [:create] do |record|
+        check_locations(record)
+        # Users cannot move a protected location
+      end
+      allow :empty_locations, [:create]
+      allow :users, [:update] do |record|
+        check_user(record, user)
+        # Users can update themselves but not change their types
+      end
       allow "api/scans", [:create]
       allow "api/coordinates", [:update]
       allow "api/locations/coordinates", [:update]
     end
+
+    private
+
+    def check_user(record, user)
+      if record.user.id == user.id && record.user.type != user.type
+        record.errors.add(:base, "You do not have permission to change your user type")
+        false
+      else
+        record.user.id == user.id
+      end
+    end
+
+    def check_locations(record)
```

Comment:
Same here. It should probably just return a boolean. Also not clear on the child_locations thing. My suggestion would be to modify the protected method in the Location model so again it just returns a boolean.